nomad/core_sched.go

```diff
@@ -706,9 +714,100 @@ func allocGCEligible(a *structs.Allocation, job *structs.Job, gcTime time.Time,
 	return timeDiff > interval.Nanoseconds()
 }
 
-// csiVolumeGC is used to garbage collect CSI volume publications
-func (c *CoreScheduler) csiVolumePublicationGC(eval *structs.Evaluation) error {
-	// TODO: implement me!
-	c.logger.Trace("garbage collecting unclaimed CSI volume publications")
+// csiVolumeClaimGC is used to garbage collect CSI volume claims
+func (c *CoreScheduler) csiVolumeClaimGC(eval *structs.Evaluation) error {
+	c.logger.Trace("garbage collecting unclaimed CSI volume claims")
+
+	// JobID smuggled in with the eval's own JobID
+	var jobID string
+	evalJobID := strings.Split(eval.JobID, ":")
+	if len(evalJobID) != 2 {
+		c.logger.Error("volume gc called without jobID")
+		return nil
+	}
+
+	jobID = evalJobID[1]
+	job, err := c.srv.State().JobByID(nil, eval.Namespace, jobID)
+	if err != nil || job == nil {
+		c.logger.Trace(
+			"cannot find job to perform volume claim GC. it may have been garbage collected",
+			"job", jobID)
+		return nil
+	}
+	c.volumeReap([]*structs.Job{job}, eval.LeaderACL)
+	return nil
+}
+
+// volumeReap contacts the leader and releases volume claims from
+// terminal allocs
+func (c *CoreScheduler) volumeReap(jobs []*structs.Job, leaderACL string) error {
+	ws := memdb.NewWatchSet()
+	for _, job := range jobs {
+		c.logger.Trace("garbage collecting unclaimed CSI volume claims for job", "job", job.ID)
+		for _, taskGroup := range job.TaskGroups {
+			for _, tgVolume := range taskGroup.Volumes {
+				if tgVolume.Type != structs.VolumeTypeCSI {
+					continue // filter to just CSI volumes
+				}
+				volID := tgVolume.Source
+				vol, err := c.srv.State().CSIVolumeByID(ws, volID)
+				if err != nil {
+					return err
+				}
+				if vol == nil {
+					c.logger.Trace("cannot find volume to be GC'd. it may have been deregistered",
+						"volume", volID)
+					return nil
+				}
+				vol, err = c.srv.State().CSIVolumeDenormalize(ws, vol)
+				if err != nil {
+					return err
+				}
+
+				gcFunc := func(alloc *structs.Allocation) error {
+					// we call denormalize on the volume above to make sure the Allocation
+					// pointer in PastAllocs isn't nil. But the alloc might have been
+					// garbage collected concurrently, so if the alloc is still nil we can
+					// safely skip it.
+					if alloc == nil || !alloc.Terminated() {
+						return nil
+					}
+					req := &structs.CSIVolumeClaimRequest{
+						VolumeID:   volID,
+						Allocation: alloc,
+						Claim:      structs.CSIVolumeClaimRelease,
+						WriteRequest: structs.WriteRequest{
+							Region:    job.Region,
+							Namespace: job.Namespace,
+						},
+					}
+					err := c.srv.controllerUnpublishVolume(req, alloc.NodeID)
+					if err != nil {
+						return err
```

Comment:
It occurs to me reading @endocrimes's comment here https://github.com/hashicorp/nomad/pull/7152/files#diff-77133a7b4c912c0ba7ead84b9d10290dR73 that we should accumulate errors here as well, rather than returning early.